### PR TITLE
[asyncio] bpo-29743: ssl socket leak during handshake

### PR DESCRIPTION
**Pull request by fafhrd91**
fixed ssl socket leak

python/asyncio#487
KeepSafe/aiohttp#1679

### CODE REVIEW

**Comment by fafhrd91** [comment]
@1st1

File: Lib/asyncio/sslproto.py [hunk 1]
```diff
@@ -569,9 +570,26 @@ def _start_handshake(self):
...
+
+    def _handshake_waiter(self, waiter):
+        if self._in_handshake and self._transport is not None:
+            if waiter.cancelled():
+                exc = base_futures.CancelledError()
+            else:
+                exc = waiter.exception()
+
+            if exc is not None:
+                self._on_handshake_complete(exc)
```
Comment (1st1) [inline Lib/asyncio/sslproto.py]: Can the `waiter` here be in a resolved state without an exception set?  I.e. is `else` clause possible here?

**Comment by fafhrd91** [comment]
I modified PR. I think aborting transport is better approach. i.e socket with uncompleted handshake is broken and we can not do anything, just drop.

**Comment by 1st1** [comment]
This is a much better version, I like it.

**Comment by Mariatta** [comment]
Is this for bpo-29743? If so, please prefix the pull request description with it. Thanks :)

**Review by 1st1 — APPROVED** [review]

File: Lib/asyncio/sslproto.py (hunk 1 above)
Comment (jdbertron) [reply Lib/asyncio/sslproto.py]: 2020 and I'm still seeing this issue when the site returns an invalid certificate. The handshake is aborted and the socket isn't released. I'm on Pyhton 3.6.8 with the latest asyncio and aiohttp.

**Pull request merged**